File: tools/src/pcr_tools/lamp_multiplex.py
"""Evidence-only multiplex LAMP planning.

DARQ/QUASR/FLOS/assimilation-style modified oligos remain method-specific.
PCRStudio does not synthesize those modifications from a standard LAMP set.
"""
from __future__ import annotations
import hashlib,json
from typing import Any
METHODS={'darq','quasr','flos','assimilation-probe','other-reviewed'}
LAMP_MULTIPLEX_SOFTWARE_MAX_TARGETS=4

def resolve(plan:Any,*,topology:str)->dict[str,Any]|None:
    if topology!='multiplex-modified-primer-probe':
        if plan not in (None,[], ''): raise ValueError('lamp_multiplex_plan requires multiplex-modified-primer-probe topology')
        return None
    if not isinstance(plan,list) or not (1<=len(plan)<LAMP_MULTIPLEX_SOFTWARE_MAX_TARGETS):
        raise ValueError('multiplex LAMP planning requires 1–3 complete peer targets (4 total including current); software bound is not wet-lab qualification')
    normalized=[]; ids=set(); reporters=set(); channels=set()
    for row in plan:
        if not isinstance(row,dict): raise ValueError('each lamp_multiplex_plan row must be an object')
        get=lambda snake,camel: str(row.get(snake) or row.get(camel) or '').strip()
        target=get('target','target'); method=get('method','method'); reporter=get('reporter','reporter'); channel=get('channel','channel'); role=get('modified_oligo_role','modifiedOligoRole'); sha=get('set_sha256','setSha256').lower(); authority=get('authority_id','authorityId'); evidence=get('empirical_evidence_ref','empiricalEvidenceRef')
        if not target or target in ids: raise ValueError('LAMP multiplex target identities must be non-empty and unique')
        ids.add(target)
        if method not in METHODS: raise ValueError('unsupported LAMP multiplex method identity')
        if not reporter or reporter in reporters or not channel or channel in channels: raise ValueError('LAMP multiplex requires unique explicit reporter/channel identities')
        reporters.add(reporter); channels.add(channel)
        if not role or not authority or not evidence: raise ValueError('LAMP multiplex requires modified oligo role, method authority and empirical evidence reference')
        if len(sha)!=64 or any(c not in '0123456789abcdef' for c in sha): raise ValueError('LAMP multiplex peer set_sha256 must be a SHA-256 hex digest')
        normalized.append({'target':target,'method':method,'reporter':reporter,'channel':channel,'modified_oligo_role':role,'set_sha256':sha,'authority_id':authority,'empirical_evidence_ref':evidence})
    digest=hashlib.sha256(json.dumps(normalized,sort_keys=True,separators=(',',':')).encode()).hexdigest()
    return {'status':'external-empirical-panel-context-complete','panel':normalized,'panel_sha256':digest,'target_count':1+len(normalized),'software_planning_bound':LAMP_MULTIPLEX_SOFTWARE_MAX_TARGETS,'wet_lab_qualified_plex':None,'automatic_modified_oligo_design':False,'sequence_decision_impact':'none','signal_claim':'empirical-target-specific-signal-required'}

File: crates/pcr-core/src/engines/lamp_multiplex.rs
//! Specialized LAMP multiplex planning/evidence boundary.
//!
//! Modified-primer/probe multiplex LAMP topologies are not a generic extension
//! of standard LAMP set design. PCRStudio records the reviewed method identity,
//! selected peer-set hash and optical/evidence context without manufacturing a
//! modified oligo or target-specific signal model.
use crate::error::{CoreError, Result};
use serde::{Deserialize, Serialize};

pub const LAMP_MULTIPLEX_SOFTWARE_MAX_TARGETS: usize = 4;
const METHODS: &[&str] = &[
    "darq",
    "quasr",
    "flos",
    "assimilation-probe",
    "other-reviewed",
];

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct LampMultiplexTarget {
    pub target: String,
    pub method: String,
    pub reporter: String,
    pub channel: String,
    pub modified_oligo_role: String,
    pub set_sha256: String,
    pub authority_id: String,
    pub empirical_evidence_ref: String,
}

pub fn validate(plan: Option<&[LampMultiplexTarget]>, topology: &str) -> Result<()> {
    if topology != "multiplex-modified-primer-probe" {
        if plan.is_some() {
            return Err(CoreError::InvalidRequest("lampMultiplexPlan is accepted only with lampDetectionTopology=`multiplex-modified-primer-probe`.".to_owned()));
        }
        return Ok(());
    }
    let plan=plan.ok_or_else(||CoreError::InvalidRequest("multiplex-modified-primer-probe requires a complete lampMultiplexPlan; PCRStudio does not infer a multiplex topology from standard LAMP primers.".to_owned()))?;
    if plan.is_empty() || plan.len() >= LAMP_MULTIPLEX_SOFTWARE_MAX_TARGETS {
        return Err(CoreError::InvalidRequest("lampMultiplexPlan must contain 1–3 peer targets (4 total including the current target); this is a software planning bound, not wet-lab qualification.".to_owned()));
    }
    let mut ids = std::collections::BTreeSet::new();
    let mut channels = std::collections::BTreeSet::new();
    let mut reporters = std::collections::BTreeSet::new();
    for row in plan {
        if row.target.trim().is_empty() || !ids.insert(row.target.trim().to_owned()) {
            return Err(CoreError::InvalidRequest(
                "lampMultiplexPlan target identities must be non-empty and unique.".to_owned(),
            ));
        }
        if !METHODS.contains(&row.method.as_str()) {
            return Err(CoreError::InvalidRequest("lampMultiplexPlan method must be darq, quasr, flos, assimilation-probe or other-reviewed.".to_owned()));
        }
        if row.reporter.trim().is_empty()
            || !reporters.insert(row.reporter.trim().to_owned())
            || row.channel.trim().is_empty()
            || !channels.insert(row.channel.trim().to_owned())
        {
            return Err(CoreError::InvalidRequest(
                "LAMP multiplex peers require unique explicit reporter and channel identities."
                    .to_owned(),
            ));
        }
        if row.modified_oligo_role.trim().is_empty()
            || row.authority_id.trim().is_empty()
            || row.empirical_evidence_ref.trim().is_empty()
        {
            return Err(CoreError::InvalidRequest("LAMP multiplex peers require modifiedOligoRole, authorityId and empiricalEvidenceRef.".to_owned()));
        }
        let sha = row.set_sha256.trim();
        if sha.len() != 64 || !sha.chars().all(|c| c.is_ascii_hexdigit()) {
            return Err(CoreError::InvalidRequest(
                "lampMultiplexPlan setSha256 must be a 64-character SHA-256 hex digest.".to_owned(),
            ));
        }
    }
    Ok(())
}
